modules/chat_utils.py

```diff
@@ -0,0 +1,194 @@
+    # zonal_url =  # point to e.g. geojson gist?
+
+    def stac_search(
+        self,
+        bbox: str,
+        dtime: str,
+        collection: str = "sentinel-2-l2a",
+        url: Optional[str] = "https://earth-search.aws.element84.com/v1/",
+    ) -> str:
+        """Perform a STAC search for Sentinel (sentinel-2-l2a) or Landsat (landsat-c2-l2) L2 images."""
+
+        self.bbox = bbox  # TODO: change to tuple?
+
+        client = Client.open(url)
+        result = client.search(collections=[collection], bbox=bbox, datetime=dtime)
+        items_dict = result.get_all_items_as_dict()
+
+        self.items_dict = items_dict
+        self.gdf = gpd.GeoDataFrame.from_features(items_dict)
+
+        return {
+            "count": result.matched(),
+        }
+
+    # TODO: switch to holoviews? or ok for demo?
+    def view_footprints(
+        self,
+    ):
+        """Load Sentinel & Landsat STAC item footprints to a map. Not for images. DO NOT use for Aqua/Terra/MODIS."""
+
+        m = (
+            self.gdf.loc[:, ["geometry"]]
+            .set_crs("epsg:4326")
+            .explore(tiles="CartoDB positron")
+        )  # TODO: could use basemap, have optional tooltips, etc.
+
+        self.media = pn.pane.plot.Folium(m, height=400)
+        return "Map is loaded to chat. Return nothing but a text confirmation to let the user know."
+
+    def plot_metadata(
+        self,
+        field: str = "eo:cloud_cover",
+    ):
+        """Plot any field from the current STAC items, e.g. cloud cover. No images, just STAC metadata fields."""
+        dates = [
+            ds.split("T")[0] for ds in self.gdf.loc[:, ["datetime"]].values.flatten()
+        ]
+        dts = [datetime.strptime(d, "%Y-%m-%d") for d in dates]
+        self.gdf.loc[:, ["date"]] = dts
+
+        self.media = pn.panel(
+                    self.gdf.loc[:, ["date", field]].set_index("date").plot()
+                )
+
+        return "Plot is loaded to chat. Return nothing other than 'Plotted!' to the user."
+
+    def set_basemap(
+        self, datestring: str = "2023-06-09", source: Optional[str] = "Aqua"
+    ):
+        """
+        Sets basemap with Modis (source = 'Aqua' or 'Terra') world coverage by date.
+        This tool does NOT require a prior STAC search. Currently only supports RGB views, no spectral indices.
+        """
+        # Valid:
+        # https://gibs.earthdata.nasa.gov/wmts/epsg4326/best/wmts.cgi
+        # ?Service=WMTS&Request=GetTile&Version=1.0.0&layer=MODIS_Terra_CorrectedReflectance_TrueColor&tilematrixset=250m
+        # &TileMatrix=6&TileCol=36&TileRow=13&TIME=2012-07-09&style=default&Format=image%2Fjpeg
+
+        tileMatrixSet = "GoogleMapsCompatible_Level9"
+        layer = f"MODIS_{source}_CorrectedReflectance_TrueColor"
+        base_url = "https://gibs.earthdata.nasa.gov/wmts/epsg3857/best/"
+        tile_path = f"{layer}/default/{datestring}/{tileMatrixSet}/" + "{Z}/{Y}/{X}.jpg"
+        self.tile_url = base_url + tile_path
```

Comment:
The url you constructed works with hv.Tiles(tile_url). However, to prevent a blank basemap you need to add the max_zoom option:

hv.Tiles(tile_url).opts(max_zoom=9)